src/d3a/models/strategy/external_strategies/storage.py

```diff
@@ -0,0 +1,261 @@
+
+    def _delete_bid_impl(self, arguments, response_channel):
+        try:
+            self.remove_bid_from_pending(arguments["bid"], self.market.id)
+            self.redis.publish_json(response_channel,
+                                    {"status": "ready", "bid_deleted": arguments["bid"]})
+        except Exception as e:
+            logging.error(f"Error when handling bid delete on area {self.device.name}: "
+                          f"Exception: {str(e)}, Bid Arguments: {arguments}")
+            self.redis.publish_json(
+                response_channel,
+                {"status": "error",
+                 "error_message": f"Error when handling bid delete "
+                                  f"on area {self.device.name} with arguments {arguments}."})
+
+    def _bid(self, payload):
+        bid_response_channel = f'{self.device.name}/bid/response'
+        if not check_for_connected_and_reply(self.redis, bid_response_channel, self.connected):
+            return
+        try:
+            arguments = json.loads(payload["data"])
+            assert set(arguments.keys()) == {'price', 'energy'}
+            arguments['buyer_origin'] = self.device.name
+        except Exception:
+            self.redis.publish_json(
+                bid_response_channel,
+                {"error": "Incorrect bid request. Available parameters: (price, energy)."}
+            )
+        else:
+            self.pending_requests.append(
+                IncomingRequest("bid", arguments, bid_response_channel))
+
+    def _bid_impl(self, arguments, bid_response_channel):
+        try:
+            self.state.clamp_energy_to_buy_kWh([self.market.time_slot])
+            max_energy = self.state.energy_to_buy_dict[self.market.time_slot]
+            assert arguments["energy"] <= max_energy
+
+            bid = self.post_bid(
+                self.market,
+                arguments["price"],
+                arguments["energy"],
+                buyer_origin=arguments["buyer_origin"]
+            )
+            self.state.offered_buy_kWh[self.market.time_slot] += arguments["energy"]
+            self.redis.publish_json(bid_response_channel,
+                                    {"status": "ready", "bid": bid.to_JSON_string()})
+        except Exception as e:
+            logging.error(f"Error when handling bid create on area {self.device.name}: "
+                          f"Exception: {str(e)}, Bid Arguments: {arguments}")
+            self.redis.publish_json(
+                bid_response_channel,
+                {"status": "error",
+                 "error_message": f"Error when handling bid create "
+                                  f"on area {self.device.name} with arguments {arguments}."})
+
+    def event_market_cycle(self):
+        self.register_on_market_cycle()
+        if self.connected:
+            self.state.market_cycle(self.market_area.current_market.time_slot,
```

Comment:
I hope that these are the only things that are needed here, and we have not missed anything else. For future reference, and as a potential improvement in the future, it would be much better to split the market_cycle event in the storage class to price/energy functions, to mimic the pattern of the pv and the load. Not required for now though